docs/android/best_practices.md

```diff
@@ -117,3 +117,89 @@ For more details, see [submit](/docs/android/submit).
 
 - **Testing**: Write [unit tests](/docs/android/testing/unit_testing) for critical functionality to ensure reliability.
 - **Code reviews**: Always review code for adherence to these best practices.
+
+## Fail fast
+
+The further you progress in development, the more difficult it becomes to debug issues. Do not ignore errors, even those you think are unlikely to occur. Always aim to catch errors at build time rather than at runtime. Use Kotlin compiler features whenever possible, and consider adding a [lint rule](/docs/android/linter) if you cannot enforce a check at compile time.
+
+### Leverage Kotlin compiler
+
+The Kotlin compiler can help you catch issues early. For example, using the `when` operator with sealed classes/interfaces ensures that all cases are handled.
+
+**Example:**
+
+```kotlin
+sealed interface Shape {
+    class Rectangle: Shape
+    class Oval: Shape
+}
+
+fun foo(shape: Shape) {
+    when(shape) {
+        is Shape.Oval -> TODO()
+        is Shape.Rectangle -> TODO()
+    }
+}
+```
+
+If you add a new class that implements `Shape`, the compiler will fail to build until you handle the new case. This is especially useful when the interface is used throughout the codebase. Note that this only works if you do not add an `else` branch.
+
+### Don't silently ignore exceptions
+
+While it is important to catch exceptions to prevent crashes, silently ignoring them can hide deeper issues and make debugging more difficult. For example, consider a third-party library that requires initialization with an API key. If initialization fails and the exception is caught without proper logging, it can be challenging to identify the root cause if something stops working in production.
+
+**Example:**
+
+```kotlin
+fun foo() {
+    
+    // Always catch the error and proceed with fallback value
+    val value = try {
+        ExternalThirdPartyJavaAPI.value()    
+    } catch (e: Exception) {
+        // Fortunately we log the error to help with troubleshooting
+        Timber.w(e,"External third party throw an error. Current state = ${ExternalThirdPartyJavaAPI.state()}")
+        "fallback"
+    }
+}
+```
+
+Proper logging ensures that users and developers can spot errors in the logs and report issues effectively.
+
+To further improve error handling during development, use the `FailFast` API. This API applies offensive programming principles by crashing the app in the `debug` flavor when an error occurs, making issues more visible early in the development process.
+
+**Example:**
+
+```kotlin
+import io.homeassistant.companion.android.common.util.FailFast
+
+fun foo() {
+    
+    // In development, this will crash the app with a message and stack trace instead of silently falling back. In production it will only print and use the fallback.
```

Comment:
```suggestion
    // In case of a failure, this will print a message and stack trace to the logs. In debug builds, it
    // will also crash the app, while in production it will use the fallback value instead of crashing.
```